mysql/resource_default_roles.go

```diff
@@ -0,0 +1,211 @@
+package mysql
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/go-version"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceDefaultRoles() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: CreateDefaultRoles,
+		UpdateContext: UpdateDefaultRoles,
+		ReadContext:   ReadDefaultRoles,
+		DeleteContext: DeleteDefaultRoles,
+		Importer: &schema.ResourceImporter{
+			StateContext: ImportDefaultRoles,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"user": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"host": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+				Default:  "localhost",
+			},
+
+			"roles": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Set: schema.HashString,
+			},
+		},
+	}
+}
+
+func checkDefaultRolesSupport(ctx context.Context, meta interface{}) error {
+	ver, _ := version.NewVersion("8.0.0")
+	if getVersionFromMeta(ctx, meta).LessThan(ver) {
+		return errors.New("MySQL version must be at least 8.0.0")
+	}
+	return nil
+}
+
+func alterUserDefaultRoles(ctx context.Context, db *sql.DB, user, host string, roles []string) error {
+	var stmtSQL string
+
+	stmtSQL = fmt.Sprintf("ALTER USER '%s'@'%s' DEFAULT ROLE ", user, host)
+
+	if len(roles) > 0 {
+		stmtSQL += fmt.Sprintf("'%s'", strings.Join(roles, "', '"))
+	} else {
+		stmtSQL += "NONE"
+	}
+
+	log.Println("Executing statement:", stmtSQL)
+	_, err := db.ExecContext(ctx, stmtSQL)
+	if err != nil {
+		return fmt.Errorf("failed executing SQL: %v", err)
```

Comment:
Low priority (if you feel like changing it here):
I know this is not used massively even in this provider (nothing depends on that), but consider using `%w` when building error from another error.
It behaves exactly like `%v` for errors, but it makes it possible to find the source error.

Reply:
Thanks for the suggestion! I changed it everywhere where it was possible. The `diag.Errorf` function does not support the `%w` format so I left `%v` in those cases.

Reply:
Thanks, that is exactly what I wanted.